README
This directory contains the Simple Python Fixed-Point Module

All files are released under the Python PSF License
(see http://www.python.org/psf/license)
and are Copyright 2006-2007 RW Penney.

The package comes with NO WARRANTY, and NO CLAIMS OF FITNESS FOR ANY PURPOSE.
However, the author welcomes *constructive* feedback
and bug-fixes via: rwpenney 'AT' users 'DOT' sourceforge 'DOT' net


## $Revision$, $Date$


ChangeLog
ChangeLog for Simple Python Fixed-Point Module

03Jun22	- *** spfpm-1.6 released

03Jun22	- Reworked decimal-precision estimation within toDecimalString()

22May22	- Added log2() method and improved efficiency for logs far from unity

09Oct21	- Migrated from distutils to setuptools

18Sep20	- Added FXfamily method for importing two's complement numbers

29Mar20	- Extended accuracy demo plots to include Machin's formula for Pi

27Mar20	- Neatened calculation of FXfamily.exp1
	Added demo plot of accuracy of exp(1)

21Mar20	- *** spfpm-1.5 released

21Mar20	- Improved testing of starting approximation within sqrt() method

15Mar20	- Improved handling of domain errors and edge-cases in __pow__() method

01Nov19	- Patched rounding behaviour on FXnum creation from float

12May18	- Added signed/twos-complement option to FXnum.toBinaryString() method

03May18	- Added 'precision' option to FXnum.toDecimalString() method
	Improved estimation of number of digits in decimal representation

28Apr18	- Replaced asin() formula for pi with Bailey-Borwein-Plouffe series

16Apr18	- *** spfpm-1.4 released

13Apr18	- Added method for printing FXnum as binary/octal/hexadecimal string

11Apr18	- Added shortcut method to FXfamily for creating FXnum instances

24Feb18	- *** spfpm-1.3 released

24Jun17	- Improved convergence of asin() series.
	Replaced atan() computation of pi with asin().

29May17	- Neatened asin() and acos() methods, avoiding use of atan().

28May17	- Improved efficiency of computation of log(2)
	Converted README.txt into markdown

13May17	- Added optimized FXnum constructor to improve speed of arithmetic
	Added command-line switches for demonstration script
	Added plot of Pi accuracy to demonstration script

06May17	- *** spfpm-1.2 released

05May17	- removed need to have same FXfamily object when combining FXnums
	- added '__slots__' specifier to FXnum fields

15Jul14	- *** spfpm-1.1 released

15Jul14	- added left/right shift operators
	improved initialization of FXnum objects from output of repr()

14Jul14	- replaced various getter methods in FXfamily with properties
	added improved expressions for computing exp(1) and log(2)
	updated graph-based demos to use Matplotlib in place of Gnuplot

19May09	- *** spfpm-1.0.1 released

20Feb09	- patched FXnum constructor to check for overflows

31Jan09	- *** spfpm-1.0 released

31Jan09	- improved accuracy of pi,log(2) etc used in trig,exp functions

06Dec08	- ported to Python-3.0

11Apr08	- *** spfpm-0.6 released

09Apr08	- added validator in FXfamily to allow limited range of integer-part

07Oct07	- added basic FLOPS calculator to demo.py

27May07	- *** spfpm-0.5 released

26May07	- minor adjustments to ensure test-cases run under Python-2.2
	added unit-test for repr()

21May07	- neatened module importing in unit-test routines

23Apr07	- removed use of 'isinstance' from FXfamily & (almost all of) FXnum

11Feb07	- added improved asin()/acos() methods based on 1-cos2t transformations

04Feb07	- removed (deprecated) FPnum wrappers

03Feb07	- *** spfpm-0.4 released

03Feb07	- patched rounding on cast to int/long to match behaviour of float-casts
	added unit-test for sqrt() method
	expanded set of exception classes, including domain-error etc
	added unit-test for asin() & acos() methods

02Feb07	- added demo/test/clean Makefile targets

01Feb07	- added setup.py distutils-script

31Jan07	- added basic asin() & acos() methods

30Jan07	- replaced _rawsincos() method with dual-mode _rawQsine() method
	added sin() & cos() methods

29Jan07	- patched spurious family-conversion in __neg__() operator

27Jan07	- extended discusions + examples in documentation
	added doctest.testmod() test for documentation
	added __abs__, __pos__, __hash__() methods
	changed default distribution-format to .zip in place of .tar.gz

22Jan07	- neatened internal usage of truncating division
	wired-together __div__ & __truediv__ operators

21Jan07	- *** spfpm-0.3 released

21Jan06	- added backwards-compatibility SetFraction method to FPnum wrapper
	renamed arctan() method to atan() to agree with math-module convention

20Jan06	- added unit-test for resolution-conversions
	added arctan & _rawarctan() methods
	adjusted sincos() method to reduce angles moduli Pi rather than halving
	added GetPi() method to FXfamily
	adjusted FXfamily to use greater bit-count when finding cached constants

19Jan06	- changed 'FPnum' to 'FXnum' to avoid abbreviation-clash with 'Floating Point'
	added FPnum backward-compatibility wrapper

18Jan06	- added unit-test for __str__ method

17Jan06	- converted resolution information to shareable FPfamily object

16Jan06	- added test for __nonzero__ method & patched reference to 'other'
	removed __iadd__, __isub__ methods to make immutable object
	added unit-test for immutability

15Jan07	- *** spfpm-0.2 released

13Jan07	- added basic Makefile for building release packages
	added unit-tests for pow() and sincos() methods

08Jan07	- added logarithm methods

03Jan07	- added basic unit-testing class for arithmetic operations

02Jan07	- converted RCS repository into SVN

26Dec06	- *** prototype of spfpm-0.1 released


graph.py
#!/usr/bin/python3
# Graphical demo of Simply Python Fixed-Point Module
# RW Penney, January 2007

import matplotlib, numpy, optparse
import matplotlib.pyplot as plt

from FixedPoint import *


def DoPlot(func, samples=range(10,26), label='function', plot_diff=False):
    fam_acc = FXfamily(32 + samples[-1])
    val_true = func(fam_acc)

    datlist = []
    for res in samples:
        fam = FXfamily(res)
        val = func(fam)

        if plot_diff:
            eps = (FXnum(val, fam_acc) - val_true)
            disc = (abs(eps)).log() / fam_acc.log2 + res
            datlist.append([res, disc, val_true])
        else:
            datlist.append([res, val, val_true])

    acc = numpy.array(datlist)

    plt.xlabel('bits')
    plt.ylabel(label)
    plt.grid(True)

    if not plot_diff:
        plt.plot(acc[:,0], acc[:,2])
    plt.plot(acc[:,0], acc[:,1])
    plt.show()


if __name__ == "__main__":
    parser = optparse.OptionParser()
    parser.add_option('-e', '--errors', default=False, action='store_true')
    opts, args = parser.parse_args()

    gconfig = [
        ( lambda fam: 4 * FXnum(1, fam). atan(),
            range(10, 26), '$4 \\tan^{-1}1$', '/tmp/graph-pi.ps' ),
        ( lambda fam: (0.5 * FXnum(2, fam).log()).exp(),
            range(5, 21), '$e^{(\\ln 2) / 2}$', '/tmp/graph-rt2.ps' )
    ]

    for cfg in gconfig:
        ( func, rng, lbl, fnm) = cfg
        if opts.errors:
            rng = range(5, 500, 10)
            lbl = 'lost-bits(' + lbl + ')'
        DoPlot(func, samples=rng, label=lbl, plot_diff=opts.errors)

# vim: set ts=4 sw=4 et:


atanpi.py
#!/usr/bin/python
# simple approximation to pi, based on atan(1/sqrt(3))
# RW Penney, December 2006

import math

terms = 15


def six():
    scl = 2 * math.sqrt(3)
    sum = 1.0
    fac = 4.0 / 9.0
    idx = 1
    while idx < terms:
        sum -= fac * (2 * idx + 1.0) / (16 * idx * idx - 1.0)
        fac /= 9.0
        idx += 1
        print(sum * scl, '   ', math.sin(sum * scl))

def eight():
    rt2 = math.sqrt(2)
    scl = 8 * math.sqrt((rt2 - 1) / (rt2 + 1))
    sum = 1.0
    fac = 1.0
    idx = 1
    while idx < terms:
        fac *= (3 - 2 * rt2) / (3 + 2 * rt2)
        sum -= fac * (8 * idx + 2 * rt2) /   \
                ((rt2 - 1.0) * (16 * idx * idx - 1.0))
        idx += 1
        print(sum * scl, '   ', math.sin(sum * scl))

def main():
    six()
    print()
    eight()

if __name__ == "__main__":
    main()


demo.py
#!/usr/bin/python3
# Demonstration of Simple Python Fixed-Point Module
# (C)Copyright 2006-2022, RW Penney

import argparse, time
from collections import OrderedDict
try:
    import matplotlib, numpy
    import matplotlib.pyplot as plt
    HAVE_MATPLOTLIB = True
except ImportError:
    HAVE_MATPLOTLIB = False

from FixedPoint import FXfamily, FXnum, FXoverflowError


def basicDemo():
    """Basic demonstration of roots & exponents at various accuracies"""

    for resolution in [8, 32, 80, 274]:
        family = FXfamily(resolution)
        val = 2

        print('=== {0} bits ==='.format(resolution))
        rt = family(val).sqrt()
        print('sqrt(' + str(val) + ')~ ' + str(rt))
        print('sqrt(' + str(val) + ')^2 ~ ' + str(rt * rt))
        print('exp(1) ~ ' + str(family.exp1))
        print()


def overflowDemo():
    """Illustrate how finite range limits calculation of exponents"""

    res = 20
    print('=== {0}-bit fractional part ==='.format(res))
    for intsize in [4, 8, 16, 32]:
        family = FXfamily(res, intsize)
        x = family(0.0)
        step = 0.1
        while True:
            try:
                ex = x.exp()
            except FXoverflowError:
                print('{0:2d}-bit integer part: exp(x) overflows near x={1:.3g}'.format(intsize, float(x)))
                break
            x += step
    print()


def speedDemo():
    """calculate indicative speed of floating-point operations"""

    print('=== speed test ===')
    for res, count in [ (16, 10000), (32, 10000),
                        (64, 10000), (128, 10000),
                        (256, 10000), (512, 10000) ]:
        fam = FXfamily(res)
        x = fam(0.5)
        lmb = fam(3.6)
        one = fam(1.0)
        t0 = time.perf_counter()
        for i in range(count):
            # use logistic-map in chaotic region:
            x = lmb * x * (one - x)
        t1 = time.perf_counter()
        ops = count * 3
        Dt = t1 - t0
        print('{0} {1}-bit arithmetic operations in {2:.2f}s ~ {3:.2g} FLOPS' \
                .format(ops, res, Dt, (ops / Dt)))

    for res, count in [ (4, 10000), (8, 10000), (12, 10000),
                        (24, 10000), (48, 10000), (128, 10000),
                        (512, 10000) ]:
        fam = FXfamily(res, 4)
        x = fam(2)
        t0 = time.perf_counter()
        for i in range(count):
            y = x.sqrt()
        t1 = time.perf_counter()
        Dt = (t1 - t0)
        print('{} {}-bit square-roots in {:.3g}s ~ {:.3g}/ms' \
                .format(count, res, Dt, count*1e-3/Dt))


def printBaseDemo():
    res = 60
    pi = FXfamily(res).pi

    print('==== Pi at {}-bit resolution ===='.format(res))
    print('decimal: {}'.format(pi.toDecimalString()))
    print('binary: {}'.format(pi.toBinaryString()))
    print('octal: {}'.format(pi.toBinaryString(3)))
    print('hex: {}'.format(pi.toBinaryString(4)))


def piPlot():
    """Plot graph of approximations to Pi"""

    b_min, b_max = 8, 25
    pi_true = FXfamily(b_max + 40).pi
    pipoints = []
    for res in range(b_min, b_max+1):
        val = 4 * FXnum(1, FXfamily(res)).atan()
        pipoints.append([res, val])
    pipoints = numpy.array(pipoints)
    truepoints = numpy.array([[b_min, pi_true], [b_max, pi_true]])

    plt.xlabel('bits')
    plt.ylabel(r'$4 \tan^{-1} 1$')
    plt.xlim([b_min, b_max])
    plt.ylim([3.13, 3.16])
    plt.grid(True)
    for arr, style in ((truepoints, '--'), (pipoints, '-')):
        plt.plot(arr[:,0], arr[:,1], ls=style)
    plt.show()


class ConstAccuracyPlot(object):
    """Plot graph of fractional bits wasted due to accumulated roundoff."""

    @classmethod
    def calcConsts(cls, fam, famOnly=False):
        """Use various methods to compute constant to given precision."""
        return ( fam.unity, FXnum(1, fam) )

    @classmethod
    def getLabels(cls):
        """Sequence of labels to use when plotting accuracy graph"""
        return ( r'1_{family}', '1' )

    @staticmethod
    def lostbits(x, x_acc):
        """Estimate of least-significant bits lost in approximation"""
        fam_acc = x_acc.family
        eps = (FXnum(x, fam_acc) - x_acc)
        return float(abs(eps).log() / fam_acc.log2
                        + x.family.resolution)

    @classmethod
    def draw(cls):
        losses = []
        for bits in range(4, 500, 4):
            fam_acc = FXfamily(bits + 40)
            fam = FXfamily(bits)
            const_true = cls.calcConsts(fam_acc, famOnly=True)[0]

            losses.append([ bits ] +
                          [ cls.lostbits(apx, const_true)
                                for apx in cls.calcConsts(fam) ])
        losses = numpy.array(losses)

        plt.xlabel('resolution bits')
        plt.ylabel('error bits')
        plt.grid(True)
        for colno, label in enumerate(cls.getLabels(), 1):
            plt.plot(losses[:,0], losses[:,colno], label=label)
        plt.legend(loc='best', fontsize='small')
        plt.show()


class PiAccuracyPlot(ConstAccuracyPlot):
    @classmethod
    def calcConsts(cls, fam, famOnly=False):
        consts = [ fam.pi ]
        if not famOnly:
            consts.append(6 * FXnum(0.5, fam).asin())
            consts.append(4 *  ( 4 * (1 / FXnum(5, fam)).atan()
                                - (1 / FXnum(239, fam)).atan() ))
        return consts

    @classmethod
    def getLabels(cls):
        return ( r'$\pi_{family}$',
                 r'$6 \sin^{-1} \frac{1}{2}$',
                 r'$\pi_{Machin}$' )


class Exp1AccuracyPlot(ConstAccuracyPlot):
    @classmethod
    def calcConsts(cls, fam, famOnly=False):
        consts = [ fam.exp1 ]
        if not famOnly:
            consts.append(1.0 / (FXnum(-0.5, fam).exp() ** 2))
        return consts

    @classmethod
    def getLabels(cls):
        return ( r'$e_{family}$',
                 r'$(e^{-1/2})^{-2}$' )


class Log2AccuracyPlot(ConstAccuracyPlot):
    @classmethod
    def calcConsts(cls, fam, famOnly=False):
        consts = [ fam.log2 ]
        if not famOnly:
            consts.append((2 * FXnum(3, fam).log()
                            - (FXnum(9, fam) / 8).log()) / 3)
        return consts

    @classmethod
    def getLabels(cls):
        return ( r'$\log2_{family}$',
                 r'$(\log9 - \log(9/8))/3$' )


def main():
    demos = OrderedDict([
        ('basic',       basicDemo),
        ('overflow',    overflowDemo),
        ('speed',       speedDemo),
        ('printing',    printBaseDemo) ])
    if HAVE_MATPLOTLIB:
        demos['piplot'] = piPlot
        demos['piaccplot'] = PiAccuracyPlot.draw
        demos['expaccplot'] = Exp1AccuracyPlot.draw
        demos['log2accplot'] = Log2AccuracyPlot.draw

    parser = argparse.ArgumentParser(
                description='Rudimentary demonstrations of'
                            ' the Simple Python Fixed-point Module (SPFPM)')
    parser.add_argument('-a', '--all', action='store_true')
    parser.add_argument('demos', nargs='*',
            help='Demo applications ({})'.format('/'.join(demos.keys())))
    args = parser.parse_args()

    if args.all or not args.demos:
        args.demos = list(demos.keys())

    for demoname in args.demos:
        demos[demoname]()


if __name__ == "__main__":
    main()

# vim: set ts=4 sw=4 et:
